(Draft Ballot) Browser Alignment (#10)
sleevi/cabforum-docs: docs/BR.md

RFC7482, Request for Comments: 7482, Registration Data Access Protocol (RDAP) Query Format, Newton, et al, March 2015.

> WebTrust for Certification Authorities, SSL Baseline with Network Security, Version 2.3, available at https://www.cpacanada.ca/-/media/site/business-and-accounting-resources/docs/webtrust/wt-pcca-ss-lbns2-3.pdf.

Reviewer: I don't think adding a direct URL (that may change in the future) or a specific version is a good idea. There was discussion with the WT representatives when we drafted language in 8.7:

> “WebTrust for CAs v2.1 or newer” AND “WebTrust for CAs SSL Baseline with Network Security v2.3 or newer”; or

Author: I think that's a separate discussion. We already have a URL, and a specific version.

#### 6.1.1.3 Subscriber Key Pair Generation
The CA SHALL reject a certificate request if the requested Public Key does not meet the requirements set forth in Sections 6.1.5 and 6.1.6 or if it has a known weak Private Key (such as a Debian weak key, see <http://wiki.debian.org/SSLkeys>).

> The CA SHALL reject a certificate request if the CA generated the Key Pair for the Subscriber and the Subscriber Certificate will contain an extendedKeyUsage extension containing either the values id-kp-serverAuth [RFC5280] or anyExtendedKeyUsage [RFC5280].

Reviewer: This sentence is a bit confusing to me (not a native English speaker). I think it is rejecting something when the CA should not generate in the first place.
Please consider the following alternative language:

> The CA SHALL NOT generate a Key Pair on behalf of a Subscriber, if the associated Subscriber Certificate will contain an extendedKeyUsage extension containing either the values id-kp-serverAuth [RFC5280] or anyExtendedKeyUsage [RFC5280].

Author: Thanks @dzacharo for raising this.

While I appreciate your attempt at the language, it introduces a big problem, and so I don't plan to adopt it, but I'm open to finding better wording. This was discussed somewhat in the servercert-wg ( https://archive.cabforum.org/pipermail/servercert-wg/2020-May/001947.html ) and the PR linked from that mail ( https://github.com/sleevi/cabforum-docs/pull/25 )

The scenario this language is trying to address is the one discussed at https://archive.cabforum.org/pipermail/servercert-wg/2020-May/001956.html

Where the CA generates a key for one usage, and the Subscriber later reuses that key for another usage. The language you propose would not address that case. This is requiring that the CA consider the keys they have *historically generated*, and not just the key they are *currently* generating, otherwise, there's a giant loophole.

Reviewer: How about:

> The CA SHALL NOT generate or use any Key Pair it generated on behalf of a Subscriber, if the associated Subscriber Certificate will contain an extendedKeyUsage extension containing either the values id-kp-serverAuth [RFC5280] or anyExtendedKeyUsage [RFC5280].

Author: ```suggestion
The CA SHALL reject a certificate request if the requested Public Key is for a Key Pair that was generated by the CA, either as part of processing this certificate request or any previous certificate request, and the Subscriber Certificate will contain an extendedKeyUsage extension containing either the values id-kp-serverAuth [RFC5280] or anyExtendedKeyUsage [RFC5280].
```

Is this any better?

Reviewer: Not really :-(

Author: Drat. My worry with your proposed language is that CAs will see "use any Key Pair it generated" as being part of processing the _current_ request, and so I want to find a way to make sure that it's as unambiguous as possible that it applies for all keys it ever generated.

This is also why "SHALL NOT generate" is, I think, a little weird, because "We didn't generate... this time", or, to use ETSI as an example, to generate a key for a non-BR use case like eSeal, and then use that key in a TLS cert. It's not the act of generation that's the problem (CAs can generate as many keys as they want), but using those keys is the problem :(

Reviewer: My first attempt was the following:

> The CA SHALL NOT generate and SHALL NOT use any Key Pair it ever generated on behalf of a Subscriber, if the associated Subscriber Certificate will contain an extendedKeyUsage extension containing either the values id-kp-serverAuth [RFC5280] or anyExtendedKeyUsage [RFC5280].

It repeats the "SHALL NOT" and has the "it **ever** generated on behalf of a Subscriber". Check it out and see if it closes the loophole.

Author: ```suggestion
If the Subscriber Certificate will contain an extendedKeyUsage extension containing either the values id-kp-serverAuth [RFC5280] or anyExtendedKeyUsage [RFC5280], the CA SHALL NOT generate a Key Pair on behalf of a Subscriber, and SHALL NOT accept a certificate request using a Key Pair previously generated by the CA. 
```

What do you think? I'm a little nervous about the "if" hanging at the end, so I restructured it a little to make it clear it applies to both clauses (generate || generated).

Reviewer: Yes, this seems much clearer than the first version. Thank you.

c. `authorityInformationAccess`

>    This extension MUST be present. It MUST NOT be marked critical, and it MUST contain the HTTP URL of the Issuing CA's OCSP responder (`accessMethod` = 1.3.6.1.5.5.7.48.1). It SHOULD also contain the HTTP URL of the Issuing CA's certificate (`accessMethod` = 1.3.6.1.5.5.7.48.2).

Reviewer: Why is this extension a mandatory requirement when it was previously allowed not to include if stapling was enforced? It seems that the current Mozilla Root Store Policy doesn't require it also (under Section 6. Revocation):

> For end-entity certificates, if the CA provides revocation information via an Online Certificate Status Protocol (OCSP) service:

Author: Could you help explain what you found confusing in the issue description, where this is addressed?

It currently reads:
> The Microsoft Root Program expresses a more tightly defined certificate profile for OCSP, and also places requirements on CRLs for intermediates. These requirements include:
>
> * OCSP MUST be supported, with no exceptions. Previously, an exception was given to contractually-defined OCSP stapling, allowing the responder URL to be omitted from the certificate. (Section 7.1.2.3 (c))

Author: The existing Microsoft program states:
> All end-entity certificates must contain an AIA extension with a valid OCSP URL. These certificates may also contain a CDP extension that contains a valid CRL URL. All other certificate types must contain either an AIA extension with an OCSP URL or a CDP extension with a valid CRL URL

Reviewer: I was under the impression that the alignment ballot wanted to capture the common policies between Root Programs, which is why I highlighted this issue. HARICA is not affected by this change but I thought that it was a good idea to exclude this requirement for Subscribers with high-volume traffic that could potentially cause problems to the CA's OCSP responders.

Author: Ah, no, you've misunderstood the purpose of this ballot. I'm definitely open for suggestions on how to make that clearer, but I tried very hard in the introduction to address this.

These are all the requirements that Root Programs impose that are not in the BRs, and which other Root Programs did not raise concerns over. The only exception I'm aware of is P-521, which is permitted by Microsoft (and they would like to continue as such) but forbidden by Mozilla and Google (and they would like it to continue as such). Beyond that, any requirement that a *single* root program has is being proposed here, because CAs are effectively already subject to this requirement, and it improves the status quo by adding to the BRs and making this auditable.

However, at least by raising this issue, you highlighted an opportunity to clarify here, which is Microsoft policy only requires one of OCSP and/or CRLs. I'll correct that.

>
> \* Non-critical Name Constraints are an exception to RFC 5280 (4.2.1.10), however, they MAY be used until the Name Constraints extension is supported by Application Software Suppliers whose software is used by a substantial portion of Relying Parties worldwide.

g. extkeyUsage (optional/required)

> For Cross Certificates that share a Subject Distinguished Name and Subject Public Key with a Root Certificate operated in accordance with these Requirements, this extension MAY be present. If present, this extension SHOULD NOT be marked critical. This extension MUST only contain usages for which the issuing CA has verified the Cross Certificate is authorized to assert. This extension MAY contain the anyExtendedKeyUsage [RFC5280] usage, if the Root Certificate(s) associated with this Cross Certificate are operated by the same organization as the issuing Root Certificate.
>
> For all other Subordinate CA Certificates, including Technically Constrained Subordinate CA Certificates:
>
> This extension MUST be present and SHOULD NOT be marked critical.
>

> > For Subordinate CA Certificates that will be used to issue TLS certificates, then either, or both, of the values id-kp-serverAuth [RFC5280] or id-kp-clientAuth [RFC5280] MUST be present. The values id-kp-emailProtection [RFC5280], id-kp-codeSigning [RFC5280], id-kp-timeStamping [RFC5280], id-kp-OCSPSigning [RFC5280], and anyExtendedKeyUsage [RFC5280] MUST NOT be present. Other values SHOULD NOT be present.

Reviewer 2: I realize this is just carrying over from the current wording, but it would seem odd to me for a subCA to have only the id-kp-clientAuth EKU, while issuing TLS certificates. I'm guessing this is a misunderstanding on my part, but would it be "better" if this read something like "For Subordinate CA Certificates that will be used to issue TLS certificates, then the value id-kp-serverAuth [RFC5280] MUST be present and the value id-kp-clientAuth [RFC5280] MAY be present."

Author: Should be fixed in https://github.com/sleevi/cabforum-docs/pull/10/commits/45208d90819a95affa7403ca87c8bf32ada0eaa7

   For Cross Certificates that share a Subject Distinguished Name and Subject Public Key with a Root Certificate operated in accordance with these Requirements, this extension MAY be present. If present, this extension SHOULD NOT be marked critical. This extension MUST only contain usages for which the issuing CA has verified the Cross Certificate is authorized to assert. This extension MAY contain the anyExtendedKeyUsage [RFC5280] usage, if the Root Certificate(s) associated with this Cross Certificate are operated by the same organization as the issuing Root Certificate.

   For all other Subordinate CA Certificates, including Technically Constrained Subordinate CA Certificates:

   This extension MUST be present and SHOULD NOT be marked critical[^**].

   For Subordinate CA Certificates that will be used to issue TLS certificates, the value id-kp-serverAuth [RFC5280] MUST be present. The value id-kp-clientAuth [RFC5280] MAY be present. The values id-kp-emailProtection [RFC5280], id-kp-codeSigning [RFC5280], id-kp-timeStamping [RFC5280], id-kp-OCSPSigning [RFC5280], and anyExtendedKeyUsage [RFC5280] MUST NOT be present. Other values SHOULD NOT be present.

>    For Subordinate CA Certificates that are not used to issue TLS certificates, then the value id-kp-serverAuth [RFC5280] MUST NOT be present. Other values MAY be present, but SHOULD NOT combine multiple independent usages (e.g. including id-kp-timeStamping [RFC5280] with id-kp-OCSPSigning [RFC5280] or id-kp-codeSigning [RFC5280]).

Reviewer: I am a bit unclear with the example of `id-kp-OCSPSigning [RFC5280]` which is probably due to my lack of in-depth knowledge on this particular EKU. Shouldn't this EKU be included along with other EKUs to allow the CA to sign delegated OCSP responder Certificates?

Author: No. This is an EKU that appears on the delegated signer, not the issuing CA. 

You can see this past discussion on m.d.s.p: https://groups.google.com/d/msg/mozilla.dev.security.policy/XQd3rNF4yOo/bXYjt1mZAwAJ

Author: I'm not sure I understand the point you're trying to make.

It's not necessary to include. The situation described was to workaround a particular CA software bug, and there are solutions to avoid that bug that do not require placing the EKU within a Sub-CA, which is not consistent with RFC 6960.

It seems like you're advocating that CAs should be able to ignore the OCSP RFC to work around a CA bug, but I could be misunderstanding?

Reviewer: I am not suggesting anything in relation to RFC 6960. My point is that your example has a SHOULD that prevents a CA from using the id-kp-OCSPSigning EKU along with -say- id-kp-codeSigning, and I think this particular EKU example is confusing. CAs should be fine including the id-kp-OCSPSigning EKU along with other EKUs. It's other EKUs that should not be allowed to co-exist with other EKUs (for example id-kp-emailProtection with id-kp-codeSigning). 
I hope this makes it somewhat clearer, at least in terms of intent.

Author: That particular EKU example is intentional.

You MUST NOT include that EKU on anything other than a delegated OCSP responder.
A delegated responder MUST NOT be capable of issuing certificates.

There's no legitimate use case for including both, except to work around a bug with a particular CA vendor, except that doing so introduces a security vulnerability, so you can and MUST work around the bug differently.

Author: See also the discussion above (e.g. https://github.com/sleevi/cabforum-docs/pull/10#issuecomment-631727791 )

c. `authorityInformationAccess`

>    This extension MUST be present. It MUST NOT be marked critical, and it MUST contain the HTTP URL of the Issuing CA's OCSP responder (`accessMethod` = 1.3.6.1.5.5.7.48.1). It SHOULD also contain the HTTP URL of the Issuing CA's certificate (`accessMethod` = 1.3.6.1.5.5.7.48.2).

Reviewer: Same comment as above. If Microsoft insists on requiring the AIA OCSP responder for all cases without exceptions, it should be clear.

Author: @dzacharo This is a very unhelpful comment. I can understand that you may feel frustrated, but this is an existing requirement of the Microsoft Root Program, which is clear there, and is clear here.

If you have a suggestion you'd like to make it clearer, I'm certainly open, but I don't think I can extract anything actionable from this comment that is useful or productive :(

Reviewer: Sure, you are the author and I am trying to review a rather large ballot. Forgive me for not always having comments you find useful. As you mentioned in [a previous comment](https://github.com/sleevi/cabforum-docs/pull/10#discussion_r440333217), Microsoft requires OCSP for end-entity certificates bur requires OCSP or CRLDP for "other cases". I don't know why the Browsers decided to follow Microsoft's Policy instead of -say- Mozilla's which seems to address the OCSP responder hammering issue I mentioned and is aligned with the existing BRs.

I have no strong feelings about this but thought it should be raised so we have a clear path, based on existing requirements and no unintentional consequences.

f. `extKeyUsage` (required)

>    Either the value `id-kp-serverAuth` [RFC5280] or `id-kp-clientAuth` [RFC5280] or both values MUST be present. id-kp-emailProtection [RFC5280] MAY be present. Other values SHOULD NOT be present. Effective 2020-07-01, the value `anyExtendedKeyUsage` MUST NOT be present.

Author: ```suggestion
   Either the value `id-kp-serverAuth` [RFC5280] or `id-kp-clientAuth` [RFC5280] or both values MUST be present. id-kp-emailProtection [RFC5280] MAY be present. Other values SHOULD NOT be present. The value `anyExtendedKeyUsage` MUST NOT be present.
```

If the signing key is P-521, the signature MUST use ECDSA with SHA-512. When encoded, the `AlgorithmIdentifier` MUST be byte-for-byte identical with the following hex-encoded bytes: `300a06082a8648ce3d040304`.

### 7.1.4 Name Forms

#### 7.1.4.1 Name Encoding
Prior to 2020-09-30, the content of the Certificate Issuer Distinguished Name field MUST match the Subject DN of the Issuing CA to support Name chaining as specified in RFC 5280, section 4.1.2.4.

Effective 2020-09-30, the following requirements SHOULD be met by all newly-issued Subordinate CA Certificates that are not used to issue TLS certificates, as defined in Section 7.1.2.2, and MUST be met for all other Certificates, regardless of whether the Certificate is a CA Certificate or a Subscriber Certificate.

For every valid Certification Path (as defined by RFC 5280, Section 6):
* For each Certificate in the Certification Path, the encoded content of the Issuer Distinguished Name field of a Certificate SHALL be byte-for-byte identical with the encoded form of the Subject Distinguished Name field of the Issuing CA certificate.

> * For each CA Certificate in the Certification Path, the encoded content of the Subject Distinguished Name field of a Certificate SHALL be byte-for-byte identical among all Certificates whose Subject Distinguished Names can be compared as equal according to RFC 5280, Section 7.1, or as modified by RFC 8399, and including expired and revoked Certificates.

Reviewer: Introducing RFC 8399. Should this also be added in section 1.6.3?

Author: Thanks. Fixed by removing, based on @CBonnell's feedback

Author: Fixed in https://github.com/sleevi/cabforum-docs/pull/10/commits/c0bfc834dee24a8f137c8699bd33274c73ec7346

## 8.5 Actions taken as a result of deficiency

## 8.6 Communication of results
The Audit Report SHALL state explicitly that it covers the relevant systems and processes used in the issuance of all Certificates that assert one or more of the policy identifiers listed in Section 7.1.6.1. The CA SHALL make the Audit Report publicly available.

The Audit Report MUST contain at least the following clearly-labelled information:
1. name of the organization being audited;
2. name and address of the organization performing the audit;

> 3. the full PKI hierarchy of all certificates that are capable of being used to issue new certificates, identified by Distinguished Name and the SHA-256 fingerprint of each and every certificate, and including all Roots, Subordinate CA Certificates, and Cross Certificates, clearly identifying which were certificates (and associated keys) were in-scope and out-of-scope of the audit;

Reviewer 3: We should decide on policy on how to demark these such that the system can look for which ones are out of scope if they are listed.

### 7.2.2 CRL and CRL entry extensions

1. `reasonCode` (OID 2.5.29.21)

  Effective 2020-09-30, the following requirements MUST be met:

  If a CRL entry is for a Root CA or Subordinate CA Certificate, including Cross Certificates, this CRL entry extension MUST be present.
  If a CRL entry is for a Certificate not technically capable of causing issuance, this CRL entry extension SHOULD be present.
  If present, this extension MUST NOT be marked critical.

>   The `CRLReason` indicated MUST NOT be unspecified (0), MUST NOT be certificateHold (6), and MUST indicate the most appropriate reason for revocation of the certificate.

Reviewer: Although this was announced in Bratislava at F2F49, this is not currently in Microsoft's Trusted Root Program. Is it expected to become part of the Microsoft Trusted Root Program before the ballot enters the discussion phase?

Author: @dzacharo I'm afraid you're mistaken that it's not part of Microsoft requirements.

You can view them at https://aka.ms/rootcert . You want Item 7 under "Communication and Disclosure Requirements"

> 7. Reason Code must be included in revocations for intermediate certificates. CAs must update the CCADB when revoking any intermediate certificates within 30 days.

This ballot reflects the technical implementation of the Microsoft requirement, along with the existing requirements within the BRs.

As such, it is part of the Root Program requirements, and it also has been reviewed by Microsoft for technical accuracy and correctness.

Reviewer: I was referring to not allowing the "unspecified" reason, which seems to be the most commonly used revocation reason and doesn't seem prohibited with the existing Microsoft Trusted Root Program requirements.

This requirement in the ballot adds more technical details and stricter language that is not in the current Microsoft set of requirements. I'm sure Microsoft has reviewed and their expectations for their policy requirements are properly captured in the proposed ballot more accurately. It just came up in my review so I wanted to flag it so CAs see what the additional expectations are.

Author: @dzacharo 

It seems like you're advocating for an interpretation that "I don't have a reason" is a valid way to fulfill the "this is the reason" requirement. That doesn't seem to at all be logical, and is intended to be prohibited. This makes it clear that it's prohibited, rather than requiring CAs be familiar with PKI technologies, which isn't always the case.

I think if you're trying to "flag it for CAs", it would be better if you paid attention to the introduction, which is where these things are flagged. Several issues seem to have already been addressed there, and so it doesn't seem productive repeating them. However, if there's something that's insufficient, I'm happy to extend that section as appropriate.
